src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -391,6 +413,8 @@ updatesWithInterleavedFlushes tr conf resolve hfs hbio root uc es reg tc = do
     let wb = tableWriteBuffer tc
         wbblobs = tableWriteBufferBlobs tc
     (wb', es') <- addWriteBufferEntries hfs resolve wbblobs maxn wb es
+    -- Supply credits before flushing, so that we complete merges in time
+    supplyCredits (V.length es - V.length es') (tableLevels tc)
```

Comment:
I think doing it only on flush is too infrequent. It would also go against the point of batching, which works best if credits accumulate in small bits up until a threshold value is met. What do you think?

Reply:
True, doing it in smaller batches might allow to spread out the work more evenly (less chance of hitting the threshold on multiple levels simultaneously). It would be good to know how much overhead it has, but it's probably negligible.

Reply:
Related: This is currently potentially doing more work than necessary, leading to a less even merging workload.

The amount credits supplied is based on the number of entries inserted. But if the key of an entry already exists in the write buffer, it won't lead to more entries being flushed. So we could base is on `size wb' - size wb`.

Reply:
@mheinzel good catch, I will change that